pkg/tuned/controller.go

```diff
@@ -411,11 +418,46 @@ func profilesExtract(profiles []tunedv1.TunedProfile) (bool, error) {
 
 		f, err := os.Create(profileFile)
 		if err != nil {
-			return change, fmt.Errorf("failed to create TuneD profile file %q: %v", profileFile, err)
+			return change, extracted, recommendedProfileDeps, fmt.Errorf("failed to create TuneD profile file %q: %v", profileFile, err)
 		}
 		defer f.Close()
 		if _, err = f.WriteString(*profile.Data); err != nil {
-			return change, fmt.Errorf("failed to write TuneD profile file %q: %v", profileFile, err)
+			return change, extracted, recommendedProfileDeps, fmt.Errorf("failed to write TuneD profile file %q: %v", profileFile, err)
+		}
+		extracted[*profile.Name] = true
+	}
+
+	return change, extracted, recommendedProfileDeps, nil
+}
+
+// profilesSync extracts TuneD daemon profiles to the daemon configuration directory
+// and removes any TuneD profiles from /etc/tuned/<profile>/ once the same TuneD
+// <profile> is no longer defined in the 'profiles' slice.
+// Returns:
+// - True if the data in the to-be-extracted recommended profile or the profiles being
+//   included from the current recommended profile have changed.
+// - Error if any or nil.
+func profilesSync(profiles []tunedv1.TunedProfile) (bool, error) {
+	change, extracted, recommendedProfileDeps, err := profilesExtract(profiles)
+	if err != nil {
+		return change, err
+	}
+
+	// Deal with TuneD profiles absent from Tuned CRs, but still present in /etc/tuned/<profile>/ the recommended profile depends on.
+	for profile := range recommendedProfileDeps {
+		if !extracted[profile] {
```

Comment:
Thanks @jmencak. Overall the code makes sense, just one idea for improvement:

I find the name of `extracted` a bit confusing here because the "problem" is that the profile IS extracted (or was), and because it was extracted, we need to delete it. However, we are checking if it is `!extracted` and if `!extracted`, we delete the corresponding profile dir.

I think renaming the variable to something like `newlyExtracted` or `extractedNew` might clarify things.

Similarly the name `recommendedProfileDeps` sounds like the list of profiles that we would need to keep in /etc/tuned/<profiles>, but it refers to the current (soon to be previous?) recommended-profile-dependencies. Maybe there is a better name we can come up with? I'm probably okay with keeping this as-is, if we make it clear that `extracted` is referring to the latest (future) needed dependencies.

Hope that makes sense, open to discussing further.

Reply:
Thank you for the suggestions, David.  I've replaced `s/extracted/extractedNew/` .  As for `recommendedProfileDeps` I do not have a better name. :/ It is one of the hardest problems in computer science. :)